src/hotspot/share/opto/output.cpp

```diff
@@ -1061,6 +1086,32 @@ void PhaseOutput::Process_OopMap_Node(MachNode *mach, int current_offset) {
       monarray->append(new MonitorValue(scval, basic_lock, eliminated));
     }
 
+    // Mark ObjectValue nodes as root nodes if they are directly
+    // referenced in the JVMS.
+    for (int i = 0; i < objs->length(); i++) {
+      ScopeValue* sv = objs->at(i);
+      if (sv->is_object_merge()) {
+        ObjectMergeValue* merge = sv->as_ObjectMergeValue();
+
+        for (int j = 0; j< merge->possible_objects()->length(); j++) {
+          ObjectValue* ov = merge->possible_objects()->at(j)->as_ObjectValue();
+          bool is_root = locarray->contains(ov) || exparray->contains(ov);
+
+          if (!is_root) {
+            for (int k = 0; k < monarray->length(); k++) {
```

Comment:
I suggest to turn the lookup over `monarray` into a helper method and call it along with `locarray` and `exparray` checks:
```
bool is_root = locarray->contains(ov) || exparray->contains(ov) || contains_as_owner(monarray, ov);
```